flang/lib/Evaluate/check-expression.cpp

```diff
@@ -135,16 +135,21 @@ bool IsConstantExprHelper<INVARIANT>::operator()(
     } else if (proc.IsPure()) {
       std::size_t j{0};
       for (const auto &arg : call.arguments()) {
-        if (const auto *dataDummy{j < proc.dummyArguments.size()
-                    ? std::get_if<characteristics::DummyDataObject>(
-                          &proc.dummyArguments[j].u)
-                    : nullptr};
-            dataDummy &&
+        const auto *dataDummy{j < proc.dummyArguments.size()
+                ? std::get_if<characteristics::DummyDataObject>(
+                      &proc.dummyArguments[j].u)
+                : nullptr};
+        if (dataDummy &&
             dataDummy->attrs.test(
                 characteristics::DummyDataObject::Attr::OnlyIntrinsicInquiry)) {
           // The value of the argument doesn't matter
         } else if (!arg) {
-          return false;
+          // Missing optional arguments are constant
+          if (!(dataDummy &&
```

Comment:
Shouldn't this be `if (!dataDummy && ...) {`?

Reply:
Nope, this code is a little weird in that it is written in such a style that it is mimic the short circuiting of the logical check. The logic is, "If the argument is missing and it isn't optional then you can short circuit and return false. Otherwise, you need to check all the other arguments (by falling through)."

I think it is a little confusing because `dataDummy` is a pointer we are checking for whether it is present. By De Morgan's dual this is equivalent to `!dataDummy || !dataDummy->attrs.test(...)`.